[doc][yba] xCluster Replication update (#23417)
yugabyte/yugabyte-db: docs/content/preview/yugabyte-platform/manage-deployments/xcluster-replication/xcluster-replication-setup.md

---
title: Set up xCluster replication for a universe
headerTitle: Set up xCluster replication
linkTitle: Setup
description: Setting up xCluster replication for a universe
headContent: Start replication from your source to your target
menu:
  preview_yugabyte-platform:
    parent: xcluster-replication
    identifier: xcluster-replication-setup
    weight: 10
type: docs
---

## Prerequisites

> Create two universes, the source universe which will serve reads and writes, and the target.

Reviewer: ```suggestion
Create the source and target universes for replication.
```

Ensure the universes have the following characteristics:

- Both universes are running the same version of YugabyteDB (v2.18.0.0 or later).

> - Both universes have the same [encryption in transit](../../../security/enable-encryption-in-transit/) settings. Encryption in transit is recommended, and you should create the source and target universes with TLS enabled.

Reviewer: ```suggestion
- Both universes have [node to node encryption](../../../security/enable-encryption-in-transit/) enabled (recommended) or both have this setting disabled.
```

The above statement seems like we require TLS for xcluster and that we require the same TLS settings (CA etc).

>     If you [rotate the CA certificate](../../../security/enable-encryption-in-transit/rotate-certificates/) on the source universe, you need to restart the replication so that the target nodes get the new root certificate for TLS verifications.

Reviewer: This should be covered in the Manage section as it is not a pre-req

    The storage configuration for copying data from source to target must be accessible from both universes. For example, if AWS S3 is used as the target of backups, the same S3 bucket must be accessible from the source and target universe.

    If an NFS file server is used, the file server or a mirror of the file server in the target region must be accessible at the same mount point on both universes. If there is significant delay in file transfer between the file server and its mirror in a different region, set the [runtime configuration](../../../administer-yugabyte-platform/manage-runtime-config/) `yb.xcluster.sleep_time_before_restore` to the maximum delay in mirroring files across the regions.

- They have enough disk space to support storage of write-ahead logs (WALs) in case of a network partition or a temporary outage of the target universe. During these cases, WALs will continue to write until replication is restored. Consider sizing your disk according to your ability to respond and recover from network or other infrastructure outages.

>     In addition, during xCluster replication setup, the source universe retains WAL logs and increases in size. The setup (including copying data from source to target) is not aborted if the source universe runs out of space. Instead, a warning is displayed notifying that the source universe has less than 100 GB of space remaining. Ensure that there is enough space available on the source universe before attempting to set up the replication. A recommended approach would be to estimate that the available disk space on the source universe is the same as the used disk space.

Reviewer: `A recommended approach would be to estimate that the available disk space on the source universe is the same as the used disk space.`

This doesn't give a good estimate in general. We can remove this line and add a better method of estimating this in the future. (cc @shahrooz1997 )

- They have network connectivity; see [Networking for xCluster](../../../prepare/networking/#networking-for-xcluster). If the source and target universe Master and TServer nodes use DNS addresses, those addresses must be resolvable on all nodes.

Prepare your database and tables on the source. The source can be empty or have data. If the source has a lot of data, setup will take longer because the [data must be copied in full to the target](#full-copy-during-xcluster-setup) before on-going asynchronous replication starts.

During initial setup, you don't need to create objects on the target. YugabyteDB Anywhere performs a full copy of the data to be replicated on the source, and automatically creates tables and objects, and restores data on the target from the source.

> For transactional YSQL, after replication is configured, the target will only be available for reads.

Reviewer: This is explained elsewhere, so we don't need to have it here. It is not a pre-req.

If a YSQL database includes tables considered ineligible for replication, this database cannot be selected in the YugabyteDB Anywhere UI and the database as a whole cannot be replicated.

You can add databases containing colocated tables to the xCluster configuration as long as the underlying database is v2.18.1.0 or later. Colocated tables on the source and target should be created with the same colocation ID if they already exist on both the source and target prior to setup. Refer to [xCluster and colocation](../../../../architecture/docdb-sharding/colocated-tables/#xcluster-and-colocation).

## Set up xCluster replication

To set up replication for a universe, do the following:

1. Navigate to your source universe and select **xCluster Replication**.

1. Click **Configure Replication**.

1. Enter a name for the replication.

1. Select the universe to use as the target.

1. Select the table type, YSQL or YCQL.

1. If you selected YSQL, choose whether to use [transactional atomicity](../#xcluster-configurations).

1. Click **Next: Select Databases** (YSQL) or **Select Tables** (YCQL).

> 1. Select the databases or tables to be copied to the target for replication.

Reviewer: This should link to the YSQL tables section. The draft doc reads

```
From a list of databases on the source universe, select the databases you want to include in the replication. See [YSQL table selection] for more details on the databases that can be selected for replication. Special attention is required for colocated databases and tables as described in that section.
```

    You can add databases containing colocated tables to the replication configuration as long as the underlying database is v2.18.1.0 or later. Colocated tables on the source and target should be created with the same colocation ID if they already exist on both the source and target prior to setup. Refer to [xCluster and colocation](../../../../architecture/docdb-sharding/colocated-tables/#xcluster-and-colocation).

1. Click **Validate Table Selection**.

    YugabyteDB Anywhere checks whether or not data needs to be copied to the target for the selected databases and its tables.

> 1. If data needs to be copied, click **Next: Configure Full Copy**, and select a storage configuration.

Reviewer: This should link to the full copy section

1. Navigate to **Admin > Alert Configurations > Alert Policies**.
1. Click **Create Alert Policy** and choose **Universe Alert**.
1. Set **Policy Template** to **Replication Lag**.
1. Enter a name and description for the alert policy.
1. Set **Target** to **Selected Universes** and select the source.
1. Set the conditions for the alert.

    - Enter a duration threshold. The alert is triggered when the lag exceeds the threshold for the specified duration.
    - Set the **Severity** option to Severe.
    - Set the **Condition** option to Greater Than.
    - Set the threshold to an acceptable value for your deployment. The default threshold is 3 minutes (180000ms).

1. Click **Save** when you are done.

When replication is set up, YugabyteDB automatically creates an alert for _YSQL Tables in DR/xCluster Config Inconsistent With Primary/Source_. This alert fires when tables are added or dropped from source's databases under replication, but are not yet added or dropped from the YugabyteDB Anywhere replication configuration.

For more information on alerting in YugabyteDB Anywhere, refer to [Alerts](../../../alerts-monitoring/alert/).

## Add a database to an existing replication

> Note that, although you don't need to create objects on the target when adding tables to an existing replication configuration, for YSQL, if a table is missing on the target, xCluster Replication must copy the entire database. By adding the objects to the target, you can potentially avoid a full copy.

Reviewer: This should link to the YSQL tables section.

YugabyteDB Anywhere proceeds to copy the database to the target. How long this takes depends mainly on the amount of data that needs to be copied.

### Full copy during xCluster setup

When xCluster is [set up](#set-up-xcluster-replication) or [restarted](#restart-replication), or when new [tables or databases are added](../xcluster-replication-ddl/) to an existing xCluster configuration, YugabyteDB Anywhere checks if a full copy is required before configuring the replication.

A full copy is required in the following circumstances:

- Any databases or tables or index tables being added are not empty. Newly created index tables associated with non-empty main tables are considered non-empty and trigger a full copy.
- Any newly added databases or tables or index tables are not present on the target universe.

For YSQL, the entire database is copied; for YCQL, the table (including associated index tables) is copied.

Ideally, you should add tables to replication right after creating them on both universes to avoid unnecessary full copy operations.

A full copy is done by first backing up the data to external storage, and then restoring to the target universe.

> - For information on creating and using storage configurations, refer to [Configure backup storage](../../../back-up-restore-universes/configure-backup-storage/)

Reviewer 2: ```suggestion
- For information on creating and using storage configurations, refer to [Configure backup storage](../../../back-up-restore-universes/configure-backup-storage/).
```

1. Click **Next: Configure Full Copy** and choose the storage configuration.
1. Click **Create a New Full Copy**.

This performs a full copy of the databases (YSQL) or tables (YCQL) involved from the source to the target.

### Pause and remove replication

To pause xCluster replication for a universe, do the following:

1. Navigate to the universe **xCluster Replication** tab and select the replication configuration.

1. Click **Pause Replication**.

To remove xCluster replication for a universe, do the following:

1. Navigate to the universe **xCluster Replication** tab and select the replication configuration.

1. Click **Actions** and choose **Delete Replication**.

You can opt to ignore errors and force delete the replication, but this is not recommended except in some rare scenarios. You can use this option to delete the xCluster configuration when one of the universes is inaccessible. However, doing so may cause some state information related to replication to be left behind on the underlying YugabyteDB universes that will require manual cleanup later on.

### Upgrading the database version

Use the same version of YugabyteDB on both the source and target.

> When [upgrading universes](../../manage-deployments/upgrade-software-install/) in xCluster replication, you should upgrade and finalize the target universe before upgrading and finalizing the source universe.

Reviewer 2: broken link

    In addition, during an outage, you will need enough disk space to retain the WALs, so determine the data change rate for your workload, and size your disk accordingly.

- [Set a replication lag alert](#set-up-replication-lag-alerts) for the source to be alerted when the replication lag exceeds acceptable levels.
- Add new tables and databases to the replication configuration soon after creating them, and before performing any writes to avoid the overhead of a full copy.
- If xCluster replication setup clashes with scheduled backups, wait for the scheduled backup to finish, and then restart the replication.

## Configure replication

### Before you start

Before configuring replication, you should be aware of the following:

- When a full copy of data needs to be made
- Special conditions for adding YSQL tables to replication

> #### Full copy during xCluster setup

Reviewer: This section and YSQL tables below could be potentially condensed or moved to the end as it currently feels like a long intro before we get to the main setup section.

    For YCQL, you can select individual tables.

1. Click **Validate Table Selection**.

    YugabyteDB Anywhere checks whether or not data needs to be copied to the target for the selected databases and its tables. See [Full copy during xCluster setup](#full-copy-during-xcluster-setup).

1. If data needs to be copied, click **Next: Configure Full Copy**, and select a storage configuration.

    The storage is used to transfer the data to the target database. For information on how to configure storage, see [Configure backup storage](../../../back-up-restore-universes/configure-backup-storage/).

1. Click **Next: Confirm Alert Threshold**.

    If you have [set an alert for replication lag](#set-up-replication-lag-alerts) on the universe, the threshold for alerting is displayed.

1. Click **Confirm and Enable Replication**.

YugabyteDB Anywhere proceeds to set up replication for the universe. How long this takes depends mainly on the amount of data that needs to be copied to the target. You can view the progress of xCluster tasks by navigating to the universe **Tasks** tab.

### Add a database to an existing replication

> Note that, although you don't need to create objects on the target when adding tables to an existing replication configuration, for YSQL, if a table is missing on the target, xCluster Replication must copy the entire database. By adding the objects to the target, you can potentially avoid a [full copy](#full-copy-during-xcluster-setup).

Reviewer: ```suggestion
Prepare your database and tables on the source. The source can be empty or have data. If the source has a lot of data, setup will take longer because the [data must be copied in full to the target](https://deploy-preview-23417--infallible-bardeen-164bc9.netlify.app/preview/yugabyte-platform/manage-deployments/xcluster-replication/xcluster-replication-setup/#full-copy-during-xcluster-setup "data must be copied in full to the target") before on-going asynchronous replication starts.

During add database, you don't need to create objects on the target. YugabyteDB Anywhere performs a full copy of the data to be replicated on the source, and automatically creates tables and objects, and restores data on the target from the source.
```

> To add a database to replication, do the following:

Reviewer: Link to the special case for bidirectional
